037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,251 @@
+
+## Motivation
+
+The rebalancing operation during scaling up and down is a manual two step process.
+
+* scaling up the cluster, then running a rebalance.
+* running a rebalance, then scaling down the cluster.
+
+In order to simplify the process, it would be great for the user to deal with just scaling the cluster by changing the `spec.kafka.replicas` field and then having the operator taking care of running a rebalancing operation automatically if requested.
+
+Taking into account the `StrimziPodSets` usage (instead of the `StatefulSets`), which will be the default in the future, a more general use case will be about adding new brokers and/or removing existing ones, not talking just about scaling up and down anymore.
+
+This proposal takes into account this future plan by referring to more general operations like add and remove brokers instead of scaling up and down.
+
+## Proposal
+
+A new `KafkaRebalanceTemplate` custom resource could be a solution allowing the user to define the configuration for the auto-rebalancing but leaving to the operator the creation of the corresponding `KafkaRebalance` custom resource.
+This new CRD should be quite similar to the `KafkaRebalance` one.
+The cluster operator has to copy over the configuration from the template to the actual rebalancing resource.
+Compared to the rejected alternative, of adding the configuration to the `Kafka` custom resource, this will avoid polluting the `Kafka` custom resource and increasing its size.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalanceTemplate
+metadata:
+  name: my-rebalance-template
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  # NOTE: mode and brokers fields are not available as in the KafkaRebalance,
+  #       they are automatically set on the corresponding KafkaRebalance by the operator
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The user has the flexibility to specify the [goals and the other rebalancing related configurations](https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaRebalance-reference) the same way they already do today on a `KafkaRebalance` custom resource.
+The cluster operator will use that configuration to fill the corresponding `KafkaRebalance` but adding the `spec.mode` and `spec.brokers` accordingly.
+It would also contain a `spec.autoApproval` field, as described by this [proposal](https://github.com/strimzi/proposals/pull/56).
+
+We would also need to extend the `Kafka` custom resource to allow the user specifying that he wants auto-rebalancing on scaling by using a template.
+A new `spec.kafka.autoRebalance` field could be added for this purpose.
```

Comment:
I still wonder if this should be part of `.spec.cruiseControl`. At the end you need to deal with that section if you want to use it.

Reply:
Yeah I would agree, in the end without CC (so the `cruiseControl` section) you cannot run a rebalance :-). I will make the change.